app/domain/authentication/authn_k8s/kubectl_exec.rb

```diff
@@ -1,133 +1,214 @@
+require 'uri'
 require 'websocket'
 require 'rubygems/package'
 
+require 'active_support/time'
+require 'websocket-client-simple'
+
+module WebSocket
+  module Client
+    module Simple
+      class Client
+        alias_method :send_msg, :send
+      end
+    end
+  end
+end
+
 module Authentication
   module AuthnK8s
-    class KubectlExec
-      STDIN_CHANNEL = 0
-      STDOUT_CHANNEL = 1
-      STDERR_CHANNEL = 2
-      ERROR_CHANNEL = 3
-      RESIZE_CHANNEL = 4
+    CommandTimedOut = ::Util::ErrorClass.new(
+      "Command timed out in container '{0}' of pod '{1}'"
+    )
+
+    class Messages
+      attr_reader :messages
+      
+      def initialize
+        @messages = Hash.new { |hash,key| hash[key] = [] }
+      end
+
+      def save_message(msg)
+        strm ||= stream_name(msg)
+        raise "Unexpected channel: #{channel(msg)}" unless strm
+        @messages[strm.to_sym] << msg_data(msg)
+      end
 
-      attr_reader :pod, :container, :timeout, :ws
+      def save_string(str, stream: nil)
```

Comment:
Should nil be the default here?  It doesn't seem to make any sense to save messages to the nil channel....

Also, this should be named clearly, perhaps (using the channel terminology suggested above):  `save_into_channel(msg)`

Reply:
Since we're only using this for errors right now, I'm going to make it `save_error_string` and just have it use the error channel.